pkg/provider/aws/resources/eks.go

```diff
@@ -65,43 +110,122 @@ type (
 
 	fargateRole := createPodExecutionRole(appName, clusterName+"-FargateExecutionRole", references)
-	dag.AddResource(fargateRole)
+	dag.AddDependenciesReflect(fargateRole)
 
 	profile := NewEksFargateProfile(cluster, subnets, fargateRole, references)
 	profile.Selectors = append(profile.Selectors, &FargateProfileSelector{Namespace: "default", Labels: map[string]string{"klotho-fargate-enabled": "true"}})
+	dag.AddDependenciesReflect(profile)
+}
 
-	dag.AddResource(profile)
-	dag.AddDependency(profile, fargateRole)
-	dag.AddDependency(profile, cluster)
+func createNodeGroups(cfg *config.Application, dag *core.ResourceGraph, units []*core.ExecutionUnit, clusterName string, cluster *EksCluster, subnets []*Subnet) []*EksNodeGroup {
+	type groupKey struct {
+		InstanceType string
+		NetworkType  string
+	}
 
-	nodeRole := createNodeRole(appName, clusterName+"-NodeGroupRole", references)
-	dag.AddResource(nodeRole)
+	type groupSpec struct {
+		DiskSizeGiB int
+		refs        []core.AnnotationKey
+	}
 
-	nodeGroup := NewEksNodeGroup(cluster, subnets, nodeRole, references)
-	dag.AddResource(nodeGroup)
-	dag.AddDependency(nodeGroup, nodeRole)
-	dag.AddDependency(nodeGroup, cluster)
+	groupSpecs := make(map[groupKey]*groupSpec)
+
+	for _, unit := range units {
+		unitCfg := cfg.GetExecutionUnit(unit.ID)
+		params := unitCfg.GetExecutionUnitParamsAsKubernetes()
+		key := groupKey{InstanceType: params.InstanceType, NetworkType: unitCfg.NetworkPlacement}
+		spec := groupSpecs[key]
+		if spec == nil {
+			spec = &groupSpec{
+				DiskSizeGiB: 20,
+			}
+			groupSpecs[key] = spec
+		}
+		spec.refs = append(spec.refs, unit.AnnotationKey)
+		if params.DiskSizeGiB > spec.DiskSizeGiB {
+			spec.DiskSizeGiB = params.DiskSizeGiB
+		}
+	}
+
+	var groups []*EksNodeGroup
+
+	hasInstanceType := false
+	for gk := range groupSpecs {
+		if gk.InstanceType != "" {
+			hasInstanceType = true
+			break
+		}
+	}
+	if !hasInstanceType {
+		for gk, spec := range groupSpecs {
+			if gk.InstanceType == "" {
+				groupSpecs[groupKey{InstanceType: "t3.medium", NetworkType: gk.NetworkType}] = spec
+			}
+		}
+	}
 
-	for _, s := range subnets {
-		dag.AddDependency(cluster, s)
-		dag.AddDependency(nodeGroup, s)
-		dag.AddDependency(profile, s)
+	for groupKey, spec := range groupSpecs {
+		if groupKey.InstanceType == "" {
+			continue
+		}
+		nodeGroup := &EksNodeGroup{
+			Name:          NodeGroupName(groupKey.NetworkType, groupKey.InstanceType),
+			ConstructsRef: spec.refs,
+			Cluster:       cluster,
+			DiskSize:      spec.DiskSizeGiB,
+			AmiType:       amiFromInstanceType(groupKey.InstanceType),
+			InstanceTypes: []string{groupKey.InstanceType},
+			Labels: map[string]string{
+				"network_placement": groupKey.NetworkType,
+			},
+			// TODO make these configurable
+			DesiredSize:    2,
+			MaxSize:        2,
+			MinSize:        1,
+			MaxUnavailable: 1,
+		}
+		nodeGroup.NodeRole = createNodeRole(cfg.AppName, fmt.Sprintf("%s.%s", clusterName, nodeGroup.Name), spec.refs)
```

Comment:
nit: i believe the cluster name already contains the app name so we might have a bit of a long redundant name here. Will still work though

Reply:
I tested this because I thought that too. `clusterName` doesn't have the app name in it, but `cluster.Name` does.